provider/postgis/util_internal_test.go

```diff
@@ -74,6 +74,128 @@ func TestReplaceTokens(t *testing.T) {
 	}
 }
 
+func TestReplaceParams(t *testing.T) {
+	type tcase struct {
+		params       map[string]provider.QueryParameterValue
+		sql          string
+		expectedSql  string
+		expectedArgs []interface{}
+	}
+
+	fn := func(tc tcase) func(t *testing.T) {
+		return func(t *testing.T) {
+			args := make([]interface{}, 0)
+			out := provider.ReplaceParams(tc.params, tc.sql, &args)
+
+			if out != tc.expectedSql {
+				t.Errorf("expected \n \t%v\n out \n \t%v", tc.expectedSql, out)
+				return
+			}
+
+			if len(tc.expectedArgs) != len(args) {
+				t.Errorf("expected \n \t%v\n out \n \t%v", tc.expectedArgs, args)
+				return
+			}
+			for i, arg := range tc.expectedArgs {
+				if arg != args[i] {
+					t.Errorf("expected \n \t%v\n out \n \t%v", tc.expectedArgs, args)
+					return
+				}
+			}
+		}
+	}
+
+	tests := map[string]tcase{
+		"nil params": {
+			params:       nil,
+			sql:          "SELECT * FROM table",
+			expectedSql:  "SELECT * FROM table",
+			expectedArgs: []interface{}{},
+		},
+		"int replacement": {
+			params: map[string]provider.QueryParameterValue{
+				"!PARAM!": {
+					Token: "!PARAM!",
+					SQL:   "?",
+					Value: 1,
+				},
+			},
+			sql:          "SELECT * FROM table WHERE PARAM = !PARAM!",
+			expectedSql:  "SELECT * FROM table WHERE PARAM = $1",
+			expectedArgs: []interface{}{1},
+		},
+		"string replacement": {
+			params: map[string]provider.QueryParameterValue{
+				"!PARAM!": {
+					Token: "!PARAM!",
+					SQL:   "?",
+					Value: "test",
+				},
+			},
+			sql:          "SELECT * FROM table WHERE PARAM = !PARAM!",
+			expectedSql:  "SELECT * FROM table WHERE PARAM = $1",
+			expectedArgs: []interface{}{"test"},
+		},
+		"null replacement": {
+			params: map[string]provider.QueryParameterValue{
+				"!PARAM!": {
+					Token: "!PARAM!",
+					SQL:   "?",
+					Value: nil,
+				},
+			},
+			sql:          "SELECT * FROM table WHERE PARAM = !PARAM!",
+			expectedSql:  "SELECT * FROM table WHERE PARAM = $1",
+			expectedArgs: []interface{}{nil},
+		},
+		"complex sql replacement": {
+			params: map[string]provider.QueryParameterValue{
+				"!PARAM!": {
+					Token: "!PARAM!",
+					SQL:   "WHERE PARAM=?",
+					Value: 1,
+				},
+			},
+			sql:          "SELECT * FROM table !PARAM!",
+			expectedSql:  "SELECT * FROM table WHERE PARAM=$1",
+			expectedArgs: []interface{}{1},
+		},
+		"subquery removal": {
+			params: map[string]provider.QueryParameterValue{
+				"!PARAM!": {
+					Token: "!PARAM!",
+					SQL:   "",
+					Value: nil,
+				},
+			},
+			sql:          "SELECT * FROM table !PARAM!",
+			expectedSql:  "SELECT * FROM table ",
+			expectedArgs: []interface{}{},
+		},
+		"multiple params": {
+			params: map[string]provider.QueryParameterValue{
+				"!PARAM!": {
+					Token: "!PARAM!",
+					SQL:   "???",
```

Comment:
Let's maybe use spaces, to at least make it look kind of correct:
```suggestion
					SQL:   "? ? ?",
```